chain-tx-enclave-next/mls/src/tree.rs

```diff
@@ -714,27 +751,46 @@ impl TreeSecret {
     ) -> Result<(), ProcessCommitError> {
         let leaf_len = tree.leaf_len();
         // Find overlap and decrypt the path secret
-        if let Some((overlap, last_secret)) =
+        if let Some((overlap, overlap_secret)) =
             self.decrypt_path_secrets(sender, tree, ctx, path_nodes, leaf_private_key)
         {
-            let mut secrets = NonEmpty::from(last_secret);
-            for _ in NodeSize::from(overlap).direct_path(leaf_len).iter() {
+            let direct_path = NodeSize::from(tree.my_pos).direct_path(leaf_len);
+            let overlap_pos = direct_path
+                .iter()
+                .position(|&p| p == overlap)
+                .expect("overlap is supposed to be ancestor");
+            let overlap_path = &direct_path[overlap_pos + 1..];
+            debug_assert_eq!(
+                overlap_path.iter().copied().collect::<Vec<_>>(),
+                NodeSize::from(overlap).direct_path(leaf_len)
+            );
+
+            // the path secrets above(not including) the overlap node
+            let mut secrets = vec![];
+            for _ in overlap_path.iter() {
                 secrets.push(tree.cs.expand_label(
-                    secrets.last(),
+                    secrets.last().unwrap_or(&overlap_secret),
                     vec![],
                     "path",
                     &[],
                     tree.cs.secret_size(),
                 )?);
             }
-            let pos = NodeSize::from(tree.my_pos)
-                .direct_path(leaf_len)
-                .iter()
-                .position(|&p| p == overlap)
-                .expect("overlap is supposed to be ancestor");
-            self.path_secrets.truncate(pos);
-            self.path_secrets
-                .extend(<Vec<SecretVec<u8>>>::from(secrets));
+
+            // verify the new path secrets match public keys
+            if !tree.verify_node_private_key(&overlap_secret, overlap) {
+                return Err(ProcessCommitError::PathSecretPublicKeyDontMatch);
+            }
+            for (secret, &parent) in secrets.iter().zip(overlap_path) {
+                if !tree.verify_node_private_key(secret, parent) {
+                    return Err(ProcessCommitError::PathSecretPublicKeyDontMatch);
+                }
+            }
+
+            self.path_secrets.truncate(overlap_pos);
```

Comment:
was a bit puzzled about `truncate` here -- I guess the meaning is that it'll keep secrets up to the overlap start (as they are unchanged), and then replace with was decrypted / computed?
a comment could be good here

Reply:
yes, added a comment